File: framework/java/src/main/java/soda/leetcode/DefaultNestedInteger.java
package soda.leetcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class DefaultNestedInteger implements NestedInteger {

    private final boolean isAtomic;

    private List<NestedInteger> list;

    private int value;

    public DefaultNestedInteger() {
        isAtomic = false;
        list = new ArrayList<>();
    }

    public DefaultNestedInteger(int value) {
        isAtomic = true;
        this.value = value;
        list = Collections.emptyList();
    }

    @Override
    public boolean isInteger() {
        return isAtomic;
    }

    @Override
    public Integer getInteger() {
        return value;
    }

    @Override
    public void setInteger(int value) {
        this.value = value;
    }

    @Override
    public void add(NestedInteger ni) {
        list.add(ni);
    }

    @Override
    public List<NestedInteger> getList() {
        return Collections.unmodifiableList(list);
    }
}

File: framework/scala/src/main/scala/soda/scala/leetcode/DefaultNestedInteger.scala
package soda.scala.leetcode

import collection.mutable

class DefaultNestedInteger(val isAtomic: Boolean) extends NestedInteger {

  private val list = mutable.ArrayBuffer[NestedInteger]()

  private var value = 0

  override def isInteger: Boolean = isAtomic

  override def getInteger: Int = value

  override def setInteger(i: Int): Unit = value = i

  override def getList: Array[NestedInteger] = list.toArray

  override def add(ni: NestedInteger): Unit = list += ni

}

object DefaultNestedInteger {
  def forInt(i: Int): DefaultNestedInteger = {
    val n = new DefaultNestedInteger(true)
    n.setInteger(i)
    n
  }
  def forList(): DefaultNestedInteger = {
    new DefaultNestedInteger(false)
  }
}
